x/staking/spec/delegation_sequence.puml

```diff
@@ -0,0 +1,104 @@
+@startuml
+'https://plantuml.com/sequence-diagram
+
+title: Delegating (currently undelegated funds delegator)
+autonumber
+
+participant "msgServer (staking)"
+participant  "keeper (staking)" as keeper
+participant validator
+participant keeper.bankKeeper
+participant vestingAccount
+participant "func:SendCoins" as SendCoins
+participant ctx.EventManager
+participant "func:DelegateCoinsFromAccountToModule" as DelegateCoinsFromAccountToModule
+participant "func:DelegateCoins" as DelegateCoins
+participant trackDelegation
+
+database store
+
+"msgServer (staking)" -> keeper : Delegate(Context, DelegatorAddress, Amount, Validator, tokenSrc := Unbonded)
+
+alt exchange rate is invalid (tokens in validator is 0)
+    keeper -->  "msgServer (staking)" : error
+end
+
+alt perform a new delegation
+    keeper -> keeper : delegation := create delegation object
+end
+
+alt new delegation
+    keeper -> keeper : BeforeDelegationCreated hook
+    note left: Calls IncrementValidatorPeriod (Used to calculate distribution) in keeper/validator.go
+end
+alt delegation exists, more tokens being added
+    keeper -> keeper : BeforeDelegationModified hook
+    note left: withdraw current delegation rewards (and increment period)
+end
+
+alt delegating from an account (subtractTokens == true)
+    keeper -> keeper.bankKeeper : DelegateCoinsFromAccountToModule
+    keeper.bankKeeper -> DelegateCoinsFromAccountToModule **
+    DelegateCoinsFromAccountToModule -->  DelegateCoinsFromAccountToModule  : Check recipient has permission to stake
+    DelegateCoinsFromAccountToModule -> DelegateCoins **
+
+    DelegateCoins --> DelegateCoins : Check the delegator has enough balances of all tokens delegated
+    DelegateCoins --> DelegateCoins : Track delegation (register that it exists to keep track of it)
+    alt validator is currently bonded
```

Comment:
Just noting that some of alts are mutually distinct choices, while some alts show a potential path based on a constraint check. So there's an opportunity to be more precise, if you think it's warranted.

Reply:
Done, using the `alt ... else` construct